llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp

```diff
@@ -7565,224 +7565,209 @@ static void reduceVSXSwap(SDNode *N, SelectionDAG *DAG) {
-    case PPC::LHA:
-    case PPC::LHA8:
-    case PPC::LHZ:
-    case PPC::LHZ8:
-    case PPC::LWZ:
-    case PPC::LWZ8:
-      FirstOp = 0;
-      break;
+  // Only additions with constant offsets will be folded.
+  if (!isa<ConstantSDNode>(MemOffset))
+    return;
+  assert(MemBase.getNumOperands() == 2 && "Invalid base of memop with offset!");
 
-    case PPC::STD:
-    case PPC::DFSTOREf64:
-    case PPC::DFSTOREf32:
-      RequiresMod4Offset = true;
-      [[fallthrough]];
-    case PPC::STB:
-    case PPC::STB8:
-    case PPC::STFD:
-    case PPC::STFS:
-    case PPC::STH:
-    case PPC::STH8:
-    case PPC::STW:
-    case PPC::STW8:
-      FirstOp = 1;
-      break;
+  auto CheckAlign = [DAG](const SDValue &Val, unsigned TargetAlign) {
+    if (GlobalAddressSDNode *GA = dyn_cast<GlobalAddressSDNode>(Val)) {
+      Align Alignment =
+          GA->getGlobal()->getPointerAlignment(DAG->getDataLayout());
+      if (Alignment < TargetAlign)
+        return false;
     }
+    return true;
+  };
 
-    // If this is a load or store with a zero offset, or within the alignment,
-    // we may be able to fold an add-immediate into the memory operation.
-    // The check against alignment is below, as it can't occur until we check
-    // the arguments to N
-    if (!isa<ConstantSDNode>(N->getOperand(FirstOp)))
-      continue;
-
-    SDValue Base = N->getOperand(FirstOp + 1);
-    if (!Base.isMachineOpcode())
-      continue;
-
-    unsigned Flags = 0;
-    bool ReplaceFlags = true;
+  SDValue ImmOp, RegOp;
+  // ADDItoc and ADDItoc8 (prints 'la') put register at second operand.
+  if (BaseOpc == PPC::ADDItoc || BaseOpc == PPC::ADDItoc8) {
+    ImmOp = MemBase.getOperand(0);
+    RegOp = MemBase.getOperand(1);
+    if (!CheckAlign(ImmOp, ExtraAlign))
+      return;
+  } else {
+    ImmOp = MemBase.getOperand(1);
+    RegOp = MemBase.getOperand(0);
+  }
+
+  // Some flags in addition needs to be carried to new memop.
+  std::optional<PPCII::TOF> NewOpFlags;
+  if (BaseOpc == PPC::ADDIdtprelL)
+    NewOpFlags = PPCII::MO_DTPREL_LO;
+  else if (BaseOpc == PPC::ADDItlsldL)
+    NewOpFlags = PPCII::MO_TLSLD_LO;
+  else if (BaseOpc == PPC::ADDItocL)
+    NewOpFlags = PPCII::MO_TOC_LO;
+
+  // On PPC64, the TOC base pointer is guaranteed by the ABI only to have
+  // 8-byte alignment, and so we can only use offsets less than 8 (otherwise,
+  // we might have needed different @ha relocation values for the offset
+  // pointers).
+  int MaxDisplacement = 7;
+  if (GlobalAddressSDNode *GA = dyn_cast<GlobalAddressSDNode>(ImmOp)) {
+    const GlobalValue *GV = GA->getGlobal();
+    Align Alignment = GV->getPointerAlignment(DAG->getDataLayout());
+    MaxDisplacement = std::min((int)Alignment.value() - 1, MaxDisplacement);
+  }
+
+  // If addis also contributes to TOC relocation, it also needs to be updated.
+  bool UpdateHaBase = false;
+  SDValue HaBase = MemBase.getOperand(0);
+
+  int Offset = cast<ConstantSDNode>(MemOffset)->getZExtValue();
```

Comment:
Can probably do `getConstantOperandVal()` here?

Reply:
Operand index of `MemOffset` depends on `IsLoad`, and we also need to check if it's a constant first.